conn.go

```diff
@@ -791,6 +795,16 @@ func marshalQueryValue(typ TypeInfo, value interface{}, dst *queryValues) error
 }
 
 func (c *Conn) executeQuery(qry *Query) *Iter {
+	ctx := qry.context
+	if rt, ok := qry.rt.(RetryPolicyWithAttemptTimeout); ok && rt.AttemptTimeout() > 0 {
+		if ctx == nil {
+			ctx = context.Background()
+		}
+		var cancel func()
+		ctx, cancel = context.WithTimeout(ctx, rt.AttemptTimeout())
```

Comment:
Timer allocation has shown to cause significant impact on query performance see #660

Reply:
Please see 81c97e428ca091d87293fa4f032d7e55a8db98d2 for my changes and some explanations. I really didn't think it would have such a noticable impact on the timing of the test but on a 25ms attempt timeout in the test each attempt is now measured as taking ~25.2ms instead of the 27-28ms previously. I took the code you wrote in #661 and copied it here because there's no good way to extract it. It works correctly (at least according to the tests) but it is somewhat convoluted.